_posts/2019-07-09-javelin.md

```diff
@@ -0,0 +1,24 @@
+---
+layout: post
+title: Javelin project
+---
+
+We are moving gfx-rs project forward, while having fun with it, thanks to the power provided by Rust language. We are building rich abstractions, which expand beyon gfx-hal API itself - into the internal layers of the backends, structured to be modular and maintainable. We are building high-performance graphics and compute infrastructure that deeply interacts with OS and drivers, thanks to the Rust's FFI capabilities and the lack of runtime. All of that, while dedicating only a bit of time and attention from the developers, who mostly do it as a side project. Rust allows us to move forward confidently and quickly, experiment with features as well as land the production code.
+
+But gfx-rs is not all written in Rust. There is one complex and important component we rely on that is written in a mixture of C and C++: [SPIRV-Cross](https://github.com/KhronosGroup/SPIRV-Cross). It's a shader translation library developed by [@TheMaister](https://github.com/HansKristian-Work) and a few Khronos members, although not an official Khronos product. We use it to generate platform-specific shaders from SPIR-V sources. It has a test suite, and its Metal backend is mostly developed and used by MoltenVK.
+
+SPIRV-Cross shows up quite a bit in our performance profiles (e.g. on Dota2). It is also written in a way (arguably!) that makes weak guarantees about what the state of the translator is at any point. And while it moves fast (in terms of contributions), it defines for us today what we can and what we can not do in terms of using advanced backend features, e.g. tessellation, argument buffers, etc. It complicates our [build process](https://github.com/gfx-rs/gfx/issues/2520), becoming a pain point for developers and users alike.
+
+We think the time has come for us [to attack](https://github.com/gfx-rs/gfx/issues/2269) that last bastion of C++ code in gfx-rs. That is what [Javelin](https://github.com/gfx-rs/javelin) project is about: "A SPIR that flies above the garden walls". It's a very complex piece of software to be written, that we haven't made much progress yet. However, we feel that Rust, once again, is the best tool for the job of shader translation: it's about parsing, working with bytes and data structures, with an ability to do unit and fuzzy testing, and no external dependencies.
```

Comment:
That is what *the* Javelin project is about...

It's a very complex piece of software and we have not made much progress on it yet.